tools/slicec-cs/src/slicec_ext/type_ref_ext.rs

```diff
@@ -3,14 +3,14 @@
 use super::{EntityExt, EnumExt, PrimitiveExt};
 use crate::cs_attributes::CsType;
 use slicec::grammar::*;
-use slicec::utils::code_gen_util::TypeContext;
 
 pub trait TypeRefExt {
     /// Is this type known to map to a C# value type?
     fn is_value_type(&self) -> bool;
 
-    /// The C# mapped type for this type reference.
-    fn cs_type_string(&self, namespace: &str, context: TypeContext, ignore_optional: bool) -> String;
+    fn field_type_string(&self, namespace: &str, ignore_optional: bool) -> String;
+    fn incoming_type_string(&self, namespace: &str, ignore_optional: bool) -> String;
+    fn outgoing_type_string(&self, namespace: &str, ignore_optional: bool) -> String;
```

Comment:
I would rename these functions incoming_parameter_type_string / outgoing_parameter_type_string

Reply:
I would add:
incoming_stream_parameter_type_string / outgoing_stream_parameter_string, where the normal functions are "no stream".

Reply:
I would replace ignore_optional by a function that removes `?` at the end of the type string, and is called when needed.